### PR TITLE
Templates for Appwrite CLI

### PR DESCRIPTION
**Pull request by christyjacob4**
This PR contains twig templates to generate the Appwrite CLI

### CODE REVIEW

File: src/SDK/Language/CLI.php [hunk 1]
```diff
@@ -0,0 +1,357 @@
+<?php
+
+namespace Appwrite\SDK\Language;
+
+use Appwrite\SDK\Language;
+
+class CLI extends Language {
```
Comment (eldadfux) [inline src/SDK/Language/CLI.php:7]: You can extend the PHP class and remove a lot of methods from this class

Comment (christyjacob4) [reply src/SDK/Language/CLI.php:7]: Perfect 👍

File: templates/cli/Dockerfile.twig [hunk 2]
```diff
@@ -0,0 +1,18 @@
+FROM composer:2.0 as step0
```
Comment (eldadfux) [inline templates/cli/Dockerfile.twig:1]: Using composer image as a production image isn't considered a best practice. We should use Docker multistage builds. You can learn more about it here: https://docs.docker.com/develop/develop-images/multistage-build/ and check the Appwrite dockerfile as a reference.

File: example.php [hunk 3]
```diff
@@ -43,253 +44,262 @@ function getSSLPage($url) {
...
-        ->setLicenseContent('test test test')
-        ->setWarning('**WORK IN PROGRESS - NOT READY FOR USAGE**')
-        ->setChangelog('**CHANGELOG**')
-        ->setVersion('0.0.1')
-        ->setGitUserName('repoowner')
-        ->setGitRepoName('reponame')
-    ;
-
-    $sdk->generate(__DIR__ . '/examples/swift');
+    // // PHP
```
Comment (eldadfux) [inline example.php:47]: We should probably refactor this code

File: templates/cli/bin/service.twig [hunk 4]
```diff
@@ -0,0 +1,3 @@
+#!/bin/sh
+
+php /usr/local/code/src/Appwrite/Services/{{service.name | caseUcfirst}}.php $@
```
Comment (eldadfux) [inline templates/cli/bin/service.twig:3]: We usually locate functional/app files (non-oop / classes) in the `app` directory and not the `src`. We would also name then in all lowercase + dashes.

File: src/SDK/Language/CLI.php [hunk 5]
```diff
@@ -0,0 +1,156 @@
...
+            ],
+            [
+                'scope'         => 'default',
+                'destination'   => '/bin/client',
+                'template'      => '/cli/bin/client.twig',
+                'minify'        => false,
+            ],
+            [
+                'scope'         => 'service',
+                'destination'   => '/src/{{ spec.title | caseUcfirst}}/Services/{{service.name | caseUcfirst}}.php',
```
Comment (eldadfux) [inline src/SDK/Language/CLI.php:143]: We usually locate functional/app files (non-oop / classes) in the `app` directory and not the `src`. We would also name then in all lowercase + dashes.

File: templates/cli/docs/example.md.twig [hunk 6]
```diff
@@ -0,0 +1,4 @@
+```sh
```
Comment (eldadfux) [inline templates/cli/docs/example.md.twig:1]: Once we have an alias we should use it for all the examples.

File: templates/cli/README.md.twig [hunk 7]
```diff
@@ -0,0 +1,30 @@
...
+{{ sdk.warning }}
+{% endif %}
+
+{{ sdk.description }}
+
+{% if sdk.logo %}
+![{{ spec.title }}]({{ sdk.logo }})
+{% endif %}
+
+## Installation
```
Comment (eldadfux) [inline templates/cli/README.md.twig:16]: We should change installation instructions once we know how we want to install the alias on users OS.

File: templates/cli/composer.json.twig [hunk 8]
```diff
@@ -0,0 +1,24 @@
...
+        "psr-4": {
+            "{{spec.title | caseUcfirst}}\\": "src/{{spec.title | caseUcfirst}}"
+        }
+    },
+    "require": {
+        "php": ">=7.4.0",
+        "utopia-php/cli": "^0.8.0",
+        "jc21/clitable": "^1.2"
+    },
+    "require-dev": {
```
Comment (eldadfux) [inline templates/cli/composer.json.twig:20]: We don't really use PHPUnit.

File: templates/cli/src/Client.php.twig [hunk 9]
```diff
@@ -0,0 +1,324 @@
+<?php
+
+namespace {{ spec.title | caseUcfirst }};
+
+use Exception;
+use Utopia\CLI\Console;
+
+class Client
+{
```
Comment (eldadfux) [inline templates/cli/src/Client.php.twig:9]: We need to make sure we support global headers just as @lohanidamodar did here: https://github.com/appwrite/sdk-generator/pull/96

File: templates/cli/src/Client.php.twig [hunk 10]
```diff
@@ -0,0 +1,324 @@
...
+        /* Load user defaults from a json file if available
+           Else Propmt the user to enter the details
+        */
+        if (!$this->loadPreferences()) {
+            $this->promptUser();
+        }
+
+        $this
+{% for header in spec.global.headers %}
+{% if header.key != 'Mode' %}
```
Comment (eldadfux) [inline templates/cli/src/Client.php.twig:54]: Very Appwrite specific, we should figure a better way to detect required security headers.

File: templates/cli/src/Client.php.twig [hunk 11]
```diff
@@ -0,0 +1,324 @@
...
+
+    /**
+     * Default User Preferences
+     *
+     * @var array
+     */
+    private $preferences = [
+        self::PREFERENCE_ENDPOINT => '',
+{% for header in spec.global.headers %}
+{% if header.key != 'Mode' %}
```
Comment (eldadfux) [inline templates/cli/src/Client.php.twig:69]: Same again, very appwrite specific.

File: templates/cli/src/Client.php.twig [hunk 12]
```diff
@@ -0,0 +1,324 @@
...
+    const METHOD_PUT = 'PUT';
+    const METHOD_PATCH = 'PATCH';
+    const METHOD_DELETE = 'DELETE';
+    const METHOD_HEAD = 'HEAD';
+    const METHOD_OPTIONS = 'OPTIONS';
+    const METHOD_CONNECT = 'CONNECT';
+    const METHOD_TRACE = 'TRACE';
+
+    const USER_PREFERENCES_FILE = __DIR__."/.prefs.json";
+    const PREFERENCE_ENDPOINT = "endpoint";
```
Comment (eldadfux) [inline templates/cli/src/Client.php.twig:21]: Is this a file path?

File: templates/cli/src/Client.php.twig [hunk 13]
```diff
@@ -0,0 +1,324 @@
...
+{% for header in spec.global.headers %}
+{% if header.key != 'Mode' %}
+        '{{header.name}}' => '',
+{% endif %}
+{% endfor %}
+    ];
+
+    public function getPreference(string $key): string
+    {
+        return $this->preferences[$key];
```
Comment (eldadfux) [inline templates/cli/src/Client.php.twig:77]: We should probably make sure key actually exist.

Comment (eldadfux) [reply templates/cli/src/Client.php.twig:77]: Maybe provide a default param & value

File: templates/cli/src/Client.php.twig [hunk 14]
```diff
@@ -0,0 +1,324 @@
...
+
+{% endfor %}
+
+    /***
+     * @param bool $status
+     * @return $this
+     */
+    public function setSelfSigned($status = true)
+    {
+        $this->selfSigned = $status;
```
Comment (eldadfux) [inline templates/cli/src/Client.php.twig:188]: Should we also store this attribute in the prefs file?

File: templates/cli/src/Parser.php.twig [hunk 15]
```diff
@@ -0,0 +1,81 @@
+<?php
+
+namespace {{ spec.title | caseUcfirst }};
+
+use jc21\CliTable;
+use jc21\CliTableManipulator;
+
+class Manipulators extends CliTableManipulator {
```
Comment (eldadfux) [inline templates/cli/src/Parser.php.twig:8]: This is non-standard in PHP. Use 1 file for 1 class.

File: templates/cli/src/Parser.php.twig [hunk 16]
```diff
@@ -0,0 +1,81 @@
...
+        'magenta',
+        'cyan',
+        'white',
+        'grey'
+    );
+
+    private const tableColor = self::colors[0];
+    private const headerColor = self::colors[2];
+
+    public function parseResponse($response, $properties = array()) {
```
Comment (eldadfux) [inline templates/cli/src/Parser.php.twig:33]: Would be nice to add doc blocks to this classes.

**Review by eldadfux — CHANGES_REQUESTED** [review]

File: templates/cli/src/Client.php.twig (hunk 12 above)
Comment (christyjacob4) [reply templates/cli/src/Client.php.twig:21]: No it is a key for a key => value pair in the preferences array. (Similar to X-Appwrite-Project etc.)

File: example.php (hunk 3 above)
Comment (christyjacob4) [reply example.php]: Done

File: templates/cli/docs/example.md.twig (hunk 6 above)
Comment (christyjacob4) [reply templates/cli/docs/example.md.twig:1]: Done

File: templates/cli/composer.json.twig (hunk 8 above)
Comment (christyjacob4) [reply templates/cli/composer.json.twig:20]: Done

File: templates/cli/Dockerfile.twig (hunk 2 above)
Comment (christyjacob4) [reply templates/cli/Dockerfile.twig:1]: As discussed, I have made use of multi-stage build

File: templates/cli/bin/service.twig (hunk 4 above)
Comment (christyjacob4) [reply templates/cli/bin/service.twig:3]: Done

File: src/SDK/Language/CLI.php (hunk 5 above)
Comment (christyjacob4) [reply src/SDK/Language/CLI.php:161]: Done

File: templates/cli/src/Parser.php.twig (hunk 15 above)
Comment (christyjacob4) [reply templates/cli/src/Parser.php.twig]: Done

File: templates/cli/src/Client.php.twig (hunk 14 above)
Comment (christyjacob4) [reply templates/cli/src/Client.php.twig]: As discussed I have set the default value to false for the CLI

File: templates/cli/src/Client.php.twig (hunk 13 above)
Comment (christyjacob4) [reply templates/cli/src/Client.php.twig:78]: Done

File: templates/cli/src/Client.php.twig (hunk 9 above)
Comment (christyjacob4) [reply templates/cli/src/Client.php.twig:9]: Done

File: templates/cli/src/Parser.php.twig (hunk 16 above)
Comment (christyjacob4) [reply templates/cli/src/Parser.php.twig]: Done.

File: templates/cli/src/Client.php.twig (hunk 10 above)
Comment (christyjacob4) [reply templates/cli/src/Client.php.twig:69]: We will retain this temporarily until it is removed from all the server-side SDK specs.

File: templates/cli/src/Client.php.twig (hunk 11 above)
Comment (christyjacob4) [reply templates/cli/src/Client.php.twig]: We will retain this temporarily until it is removed from all the server-side SDK specs.

File: templates/cli/README.md.twig (hunk 7 above)
Comment (christyjacob4) [reply templates/cli/README.md.twig:16]: Done

**Review by eldadfux — APPROVED** [review]

**Pull request merged**